pkg/tls/tls.go

```diff
@@ -45,20 +46,101 @@ func prepareCommonConfig(peerCert string, peerKey string) *tls.Config {
+	certLoader, err := NewPeriodicCertLoader(peerCert, peerKey, syncPeriod)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"err": err,
+		}).Fatalln("Error starting TLS certificate reloader")
+	}
+	go certLoader.Start()
+
+	config := &tls.Config{}
+	config.GetCertificate = func(*tls.ClientHelloInfo) (*tls.Certificate, error) {
+		return certLoader.Current(), nil
+	}
+	return config
+}
+
+func prepareClientConfigWithPeriodicReload(peerCert string, peerKey string, syncPeriod time.Duration) *tls.Config {
+	certLoader, err := NewPeriodicCertLoader(peerCert, peerKey, syncPeriod)
+	if err != nil {
+		log.WithFields(log.Fields{
+			"err": err,
+		}).Fatalln("Error starting TLS certificate reloader")
+	}
+	go certLoader.Start()
+
+	config := &tls.Config{}
+	config.GetClientCertificate = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		return certLoader.Current(), nil
+	}
+	return config
+}
+
+// PrepareServerConfig prepares a TLS config instance for a server that wants mTLS enabled.
+// DEPRECATED: use PrepareMutualTLSServerConfig instead
 func PrepareServerConfig(rootCACerts []string, peerCert string, peerKey string) *tls.Config {
+	return PrepareMutualTLSServerConfig(rootCACerts, peerCert, peerKey)
+}
+
+// PrepareMutualTLSServerConfig prepares a TLS config instance for a server that wants mTLS enabled.
+func PrepareMutualTLSServerConfig(rootCACerts []string, peerCert string, peerKey string) *tls.Config {
 	config := prepareCommonConfig(peerCert, peerKey)
 	config.ClientCAs = prepareCertPool(rootCACerts...)
 	config.ClientAuth = tls.RequireAndVerifyClientCert
 	return config
 }
 
-// PrepareClientConfig prepares a TLS config instance for a client that wants
-// mTLS enabled.
+// PrepareMutualTLSServerConfigPeriodicWithReload prepares a TLS config instance for a server that wants mTLS enabled
+// and dynamic reloading of server certificates.
+func PrepareMutualTLSServerConfigWithPeriodicReload(rootCACerts []string, peerCert string, peerKey string, syncPeriod time.Duration) *tls.Config {
+	config := prepareServerConfigWithPeriodicReload(peerCert, peerKey, syncPeriod)
+	config.ClientCAs = prepareCertPool(rootCACerts...)
+	config.ClientAuth = tls.RequireAndVerifyClientCert
+	return config
+}
+
+// PrepareTLSServerConfig prepares a TLS config for a server with regular TLS
+func PrepareTLSServerConfig(peerCert string, peerKey string) *tls.Config {
+	config := prepareCommonConfig(peerCert, peerKey)
+	return config
+}
+
+// PrepareTLSServerConfigWithWatcher prepares a TLS config for a server with regular TLS
+// and dynamic reloading of server certificates.
+func PrepareTLSServerConfigWithPeriodicReload(peerCert string, peerKey string, syncPeriod time.Duration) *tls.Config {
+	config := prepareServerConfigWithPeriodicReload(peerCert, peerKey, syncPeriod)
+	return config
+}
+
+// PrepareClientConfig prepares a TLS config instance for a client that wants mTLS enabled.
```

Comment:
@andrewsykim I think that for the sake of clarity you should remove this func instead of calling the new one.